File: CS/ViewModels/ContactsViewModel.cs
namespace DXHtmlMessengerSample.ViewModels {
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using DevExpress.DevAV.Chat;
    using DevExpress.DevAV.Chat.Commands;
    using DevExpress.DevAV.Chat.Events;
    using DevExpress.DevAV.Chat.Model;
    using DevExpress.Mvvm;
    using DevExpress.Mvvm.DataAnnotations;
    using DevExpress.Mvvm.POCO;

    public class ContactsViewModel : ChannelViewModel {
        public ContactsViewModel()
            : base() {
            Contacts = new Contact[0];
            Messenger.Default.Register<Contact>(this, OnContact);
        }
        protected override void OnConnected(IChannel channel) {
            base.OnConnected(channel);
            channel.Subscribe(OnContactEvents);
        }
        protected override async void OnChannelReady() {
            var channelContacts = await Channel.GetContacts();
            await DispatcherService?.BeginInvoke(() => Contacts = channelContacts);
        }
        void OnContact(Contact contact) {
            UpdateSelectedContact(contact);
        }
        async void OnContactEvents(Dictionary<long, ContactEvent> events) {
            ContactEvent @event;
            foreach(Contact contact in Contacts) {
                if(events.TryGetValue(contact.ID, out @event))
                    @event.Apply(contact);
            }
            if(events.Count > 0)
                await DispatcherService?.BeginInvoke(RaiseContactsChanged);
        }
        void RaiseContactsChanged() {
            this.RaisePropertyChanged(x => x.Contacts);
        }
        public virtual IReadOnlyCollection<Contact> Contacts {
            get;
            protected set;
        }
        protected void OnContactsChanged() {
            if(SelectedContact == null)
                SelectedContact = Contacts.FirstOrDefault();
            else UpdateSelectedContact(SelectedContact);
            }
        public virtual Contact SelectedContact {
            get;
            set;
        }
        protected void OnSelectedContactChanged() {
            NotifyContactSelected(SelectedContact);
            this.RaiseCanExecuteChanged(x => x.ClearConversation());
            this.RaiseCanExecuteChanged(x => x.CopyContact());
        }
        int lockContact = 0;
        void UpdateSelectedContact(Contact contact) {
            if(lockContact > 0 || contact == null)
                return;
            lockContact++;
            try {
                long id = contact.ID;
                SelectedContact = Contacts.Where(x => x.ID == id).FirstOrDefault() ?? Contacts.FirstOrDefault();
            }
            finally { lockContact--; }
        }
        void NotifyContactSelected(Contact contact) {
            if(lockContact > 0 || contact == null)
                return;
            lockContact++;
            try {
                Messenger.Default.Send(contact);
            }
            finally { lockContact--; }
        }
        public async void ShowContact(Contact contact) {
            var contactInfo = await Channel.GetUserInfo(contact.ID);
            MessengerViewModel.ShowContactInfo(this, contactInfo);
        }
        [Command(false)]
        public bool HasContact() {
            return SelectedContact != null;
        }
        [Command(CanExecuteMethodName = nameof(HasContact))]
        public void ClearConversation() {
            if(Channel != null)
                Channel.Send(new ClearConversation(SelectedContact));
        }
        [Command(CanExecuteMethodName = nameof(HasContact))]
        public async void CopyContact() {
            try {
                var info = await Channel.GetUserInfo(SelectedContact.ID);
                string contact =
                    info.Name + System.Environment.NewLine +
                    $"Email: {info.Email}" + System.Environment.NewLine +
                    $"Phone: {info.MobilePhone}";
                System.Windows.Forms.Clipboard.SetText(contact);
            }
            catch { }
        }
        ContactViewModel contactTooltipViewModel;
        public async Task<ContactViewModel> EnsureTooltipViewModel(Contact contact) {
            var contactInfo = await Channel.GetUserInfo(contact.ID);
            if(contactTooltipViewModel == null)
                contactTooltipViewModel = ContactViewModel.Create(contactInfo);
            else
            ((ISupportParameter)contactTooltipViewModel).Parameter = contactInfo;
            return contactTooltipViewModel;
        }
    }
}

File: VB/ViewModels/ContactsViewModel.vb
Imports System.Threading.Tasks
Imports DevExpress.DevAV.Chat
Imports DevExpress.DevAV.Chat.Commands
Imports DevExpress.DevAV.Chat.Events
Imports DevExpress.DevAV.Chat.Model
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.POCO

Namespace DXHtmlMessengerSample.ViewModels
    Public Class ContactsViewModel
        Inherits ChannelViewModel
        Public Sub New()
            MyBase.New()
            Contacts = New Contact() {}
            Messenger.Default.Register(Of Contact)(Me, AddressOf OnContact)
        End Sub
        Protected Overrides Sub OnConnected(ByVal channel As IChannel)
            MyBase.OnConnected(channel)
            channel.Subscribe(AddressOf OnContactEvents)
        End Sub
        Protected Overrides Async Sub OnChannelReady()
            Dim channelContacts = Await Channel.GetContacts()
            Await DispatcherService?.BeginInvoke(Sub() Contacts = channelContacts)
        End Sub
        Sub OnContact(ByVal contact As Contact)
            UpdateSelectedContact(contact)
        End Sub
        Async Sub OnContactEvents(ByVal events As Dictionary(Of Long, ContactEvent))
            Dim [event] As ContactEvent = Nothing
            For Each contact As Contact In Contacts
                If events.TryGetValue(contact.ID, [event]) Then
                    [event].Apply(contact)
                End If
            Next contact
            If events.Count > 0 Then
                Await DispatcherService?.BeginInvoke(AddressOf RaiseContactsChanged)
            End If
        End Sub
        Sub RaiseContactsChanged()
            Me.RaisePropertyChanged(Function(x) x.Contacts)
        End Sub
        Public Overridable Property Contacts() As IReadOnlyCollection(Of Contact)
        Protected Sub OnContactsChanged()
            If SelectedContact Is Nothing Then
                SelectedContact = Contacts.FirstOrDefault()
            Else
                UpdateSelectedContact(SelectedContact)
            End If
        End Sub
        Public Overridable Property SelectedContact() As Contact
        Protected Sub OnSelectedContactChanged()
            NotifyContactSelected(SelectedContact)
            Me.RaiseCanExecuteChanged(Sub(x) x.ClearConversation())
            Me.RaiseCanExecuteChanged(Sub(x) x.CopyContact())
        End Sub
        Private lockContact As Integer = 0
        Sub UpdateSelectedContact(ByVal contact As Contact)
            If lockContact > 0 OrElse contact Is Nothing Then
                Return
            End If
            lockContact += 1
            Try
                Dim id As Long = contact.ID
                SelectedContact = If(Contacts.Where(Function(x) x.ID = id).FirstOrDefault(), Contacts.FirstOrDefault())
            Finally
                lockContact -= 1
            End Try
        End Sub
        Sub NotifyContactSelected(ByVal contact As Contact)
            If lockContact > 0 OrElse contact Is Nothing Then
                Return
            End If
            lockContact += 1
            Try
                Messenger.Default.Send(contact)
            Finally
                lockContact -= 1
            End Try
        End Sub
        Public Async Sub ShowContact(ByVal contact As Contact)
            Dim contactInfo = Await Channel.GetUserInfo(contact.ID)
            MessengerViewModel.ShowContactInfo(Me, contactInfo)
        End Sub
        <Command(False)>
        Public Function HasContact() As Boolean
            Return SelectedContact IsNot Nothing
        End Function
        <Command(CanExecuteMethodName:=NameOf(HasContact))>
        Public Sub ClearConversation()
            If Channel IsNot Nothing Then
                Channel.Send(New ClearConversation(SelectedContact))
            End If
        End Sub
        <Command(CanExecuteMethodName:=NameOf(HasContact))>
        Public Async Sub CopyContact()
            Try
                Dim info = Await Channel.GetUserInfo(SelectedContact.ID)
                Dim contact As String = info.Name & System.Environment.NewLine & $"Email: {info.Email}" & System.Environment.NewLine & $"Phone: {info.MobilePhone}"
                System.Windows.Forms.Clipboard.SetText(contact)
            Catch
            End Try
        End Sub
        Private contactTooltipViewModel As ContactViewModel
        Public Async Function EnsureTooltipViewModel(ByVal contact As Contact) As Task(Of ContactViewModel)
            Dim contactInfo = Await Channel.GetUserInfo(contact.ID)
            If contactTooltipViewModel Is Nothing Then
                contactTooltipViewModel = ContactViewModel.Create(contactInfo)
            Else
                DirectCast(contactTooltipViewModel, ISupportParameter).Parameter = contactInfo
            End If
            Return contactTooltipViewModel
        End Function
    End Class
End Namespace